### ISSUE TITLE
armv6l: undefined reference to `__atomic_store_8'

### ISSUE THREAD

**Issue by amak79** [badaix/snapcast#490]
Attempting to build Snapcast 0.16.0 on Void Linux armv6l results in the following error:
```
[ 64%] Building CXX object client/CMakeFiles/snapclient.dir/decoder/pcm_decoder.cpp.o
[ 71%] Building CXX object client/CMakeFiles/snapclient.dir/player/player.cpp.o
[ 78%] Building CXX object client/CMakeFiles/snapclient.dir/player/alsa_player.cpp.o
[ 85%] Building CXX object client/CMakeFiles/snapclient.dir/decoder/ogg_decoder.cpp.o
[ 92%] Building CXX object client/CMakeFiles/snapclient.dir/decoder/flac_decoder.cpp.o
[100%] Linking CXX executable ../../bin/snapclient
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: CMakeFiles/snapclient.dir/controller.cpp.o: in function `Controller::worker()':
controller.cpp:(.text+0x25c0): undefined reference to `__atomic_load_8'
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: CMakeFiles/snapclient.dir/stream.cpp.o: in function `Stream::getPlayerChunk(void*, std::chrono::duration<long long, std::ratio<1ll, 1000000ll> > const&, unsigned long)':
stream.cpp:(.text+0x2e78): undefined reference to `__atomic_load_8'
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: stream.cpp:(.text+0x321c): undefined reference to `__atomic_load_8'
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: stream.cpp:(.text+0x433c): undefined reference to `__atomic_load_8'
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: stream.cpp:(.text+0x53c8): undefined reference to `__atomic_load_8'
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: CMakeFiles/snapclient.dir/time_provider.cpp.o: in function `TimeProvider::setDiffToServer(double)':
time_provider.cpp:(.text+0xb20): undefined reference to `__atomic_store_8'
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: time_provider.cpp:(.text+0xed0): undefined reference to `__atomic_store_8'
collect2: error: ld returned 1 exit status
make[2]: *** [client/CMakeFiles/snapclient.dir/build.make:220: ../bin/snapclient] Error 1
make[1]: *** [CMakeFiles/Makefile2:145: client/CMakeFiles/snapclient.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
The complete build log is available on the Void Linux CI: https://travis-ci.org/void-linux/void-packages/builds/602121587

**Comment by badaix** [comment]
should be fixed with 2eec944
can you try if it's working now?

**Comment by amak79** [comment]
Same error.

**Comment by badaix** [comment]
once again please c4a8795

**Comment by amak79** [comment]
Fails with the same error, but cmake now checks for `__atomic_load_8`:
```
-- Looking for __atomic_fetch_add_4 in atomic
-- Looking for __atomic_fetch_add_4 in atomic - not found
-- Looking for __atomic_load_8 in atomic
-- Looking for __atomic_load_8 in atomic - not found
```

**Comment by amak79** [comment]
I was missing the `libatomic-devel` dep when building armv6l. Adding that dep  and setting `LDFLAGS=" -latomic"` allows the build to complete successfully. However even with that dep installed, cmake still reports `__atomic_fetch_add_4` and `__atomic_load_8` as not found.

**Comment by amak79** [comment]
I realized that adding the `libatomic-devel` dep is not enough to fix this issue. I had explicitly set `HAS_LIBATOMIC` during testing. That's why the build succeeded when I added the `libatomic-devel` dep. The build fails when `HAS_LIBATOMIC` hasn't been explicitly set, so there is still an issue with atomic detection.

**Comment by badaix** [comment]
The third parameter to `check_library_exists` is the search path. I wonder if this is the issue on your system.

**Comment by amak79** [comment]
The Void Linux package build system (`xbps-src`) installs the `libatomic-devel` files in the following location:
```
~/void-packages/masterdir/usr/arm-linux-gnueabihf/usr/lib/libatomic.a
~/void-packages/masterdir/usr/arm-linux-gnueabihf/usr/lib/libatomic.so -> libatomic.so.1.2.0
~/void-packages/masterdir/usr/arm-linux-gnueabihf/usr/lib/libatomic.so.1 -> libatomic.so.1.2.0
~/void-packages/masterdir/usr/arm-linux-gnueabihf/usr/lib/libatomic.so.1.2.0
```
During the build `xbps-src` sets up a chroot at `~/void-packages/masterdir/usr/arm-linux-gnueabihf`. So effectively the files are located in `/usr/lib`.

Both `__atomic_fetch_add_4` and `__atomic_load_8` exist:
```
$ nm ~/void-packages/masterdir/usr/arm-linux-gnueabihf/usr/lib/libatomic.a | grep -E '__atomic_fetch_add_4|__atomic_load_8'
         U __atomic_load_8
00000088 i __atomic_fetch_add_4
00000024 i __atomic_load_8
```
I also tried setting the `check_library_exists` search path to `/usr/lib` but cmake still can't find libatomic.

**Comment by amak79** [comment]
CMakeError.txt
```
Determining if the function __atomic_fetch_add_4 exists in the atomic failed with the following output:
Change Dir: /builddir/snapcast-0.16.0/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_6692d/fast && /usr/bin/make -f CMakeFiles/cmTC_6692d.dir/build.make CMakeFiles/cmTC_6692d.dir/build
make[1]: Entering directory '/builddir/snapcast-0.16.0/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_6692d.dir/CheckFunctionExists.cxx.o
/builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++    -fstack-clash-protection -D_FORTIFY_SOURCE=2 -O2 -pipe -march=armv6 -mfpu=vfp -mfloat-abi=hard   -I/usr/arm-linux-gnueabihf/usr/include -DCHECK_FUNCTION_EXISTS=__atomic_fetch_add_4   -o CMakeFiles/cmTC_6692d.dir/CheckFunctionExists.cxx.o -c /builddir/snapcast-0.16.0/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx
<command-line>: error: new declaration 'char __atomic_fetch_add_4()' ambiguates built-in declaration 'unsigned int __atomic_fetch_add_4(volatile void*, unsigned int, int)' [-fpermissive]
<command-line>: note: in definition of macro 'CHECK_FUNCTION_EXISTS'
/builddir/snapcast-0.16.0/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx: In function 'int main(int, char**)':
/builddir/snapcast-0.16.0/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx:17:25: error: too few arguments to function 'unsigned int __atomic_fetch_add_4(volatile void*, unsigned int, int)'
   17 |   CHECK_FUNCTION_EXISTS();
      |                         ^
make[1]: *** [CMakeFiles/cmTC_6692d.dir/build.make:66: CMakeFiles/cmTC_6692d.dir/CheckFunctionExists.cxx.o] Error 1
make[1]: Leaving directory '/builddir/snapcast-0.16.0/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_6692d/fast] Error 2



Determining if the function __atomic_load_8 exists in the atomic failed with the following output:
Change Dir: /builddir/snapcast-0.16.0/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_d183f/fast && /usr/bin/make -f CMakeFiles/cmTC_d183f.dir/build.make CMakeFiles/cmTC_d183f.dir/build
make[1]: Entering directory '/builddir/snapcast-0.16.0/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_d183f.dir/CheckFunctionExists.cxx.o
/builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++    -fstack-clash-protection -D_FORTIFY_SOURCE=2 -O2 -pipe -march=armv6 -mfpu=vfp -mfloat-abi=hard   -I/usr/arm-linux-gnueabihf/usr/include -DCHECK_FUNCTION_EXISTS=__atomic_load_8   -o CMakeFiles/cmTC_d183f.dir/CheckFunctionExists.cxx.o -c /builddir/snapcast-0.16.0/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx
<command-line>: error: new declaration 'char __atomic_load_8()' ambiguates built-in declaration 'long long unsigned int __atomic_load_8(const volatile void*, int)' [-fpermissive]
<command-line>: note: in definition of macro 'CHECK_FUNCTION_EXISTS'
/builddir/snapcast-0.16.0/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx: In function 'int main(int, char**)':
/builddir/snapcast-0.16.0/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx:17:25: error: too few arguments to function 'long long unsigned int __atomic_load_8(const volatile void*, int)'
   17 |   CHECK_FUNCTION_EXISTS();
      |                         ^
make[1]: *** [CMakeFiles/cmTC_d183f.dir/build.make:66: CMakeFiles/cmTC_d183f.dir/CheckFunctionExists.cxx.o] Error 1
make[1]: Leaving directory '/builddir/snapcast-0.16.0/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_d183f/fast] Error 2
```
This issue seems to be the same as the one reported here: https://github.com/aws/aws-sdk-cpp/issues/1199

**Comment by amak79** [comment]
The following patch seems to correctly detect atomic operations. I adapted it from from this pull request: https://github.com/aws/aws-sdk-cpp/pull/1243 
```
diff -urpN a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2019-10-24 21:53:43.000000000 +1100
+++ b/CMakeLists.txt	2019-10-28 20:52:11.418575657 +1100
@@ -77,11 +77,38 @@ target_architecture(ARCH)
 #message(STATUS "Architecture: ${ARCH}")
 #message(STATUS "System processor: ${CMAKE_SYSTEM_PROCESSOR}")
 
-INCLUDE(CheckLibraryExists)
-check_library_exists(atomic __atomic_fetch_add_4 "" HAS_LIBATOMIC)
-if(HAS_LIBATOMIC)
-    set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -latomic")
+set(ATOMIC_TEST_CXX_SOURCE "
+#include <cstdint>
+#include <atomic>
+std::atomic<uintptr_t> x;
+std::atomic<uintmax_t> y;
+int main() {
+    return x + y;
+}")
+
+set(LIBATOMIC_TEST_CXX_SOURCE "
+int main() {
+    __atomic_load_8(nullptr, 0);
+    return 0;
+}")
+
+include(CMakePushCheckState)
+include(CheckCXXSourceCompiles)
+cmake_push_check_state()
+check_cxx_source_compiles("${ATOMIC_TEST_CXX_SOURCE}" HAVE_ATOMICS_WITHOUT_LIBATOMIC)
+if(NOT HAVE_ATOMICS_WITHOUT_LIBATOMIC)
+    set(CMAKE_REQUIRED_LIBRARIES atomic)
+    check_cxx_source_compiles("${LIBATOMIC_TEST_CXX_SOURCE}" LIBATOMIC_EXISTS)
+    if(LIBATOMIC_EXISTS)
+        check_cxx_source_compiles("${ATOMIC_TEST_CXX_SOURCE}" HAVE_ATOMICS_WITH_LIBATOMIC)
+    endif()
+    if(HAVE_ATOMICS_WITH_LIBATOMIC)
+        set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -latomic")
+    else()
+        message(FATAL_ERROR "Could not determine support for atomic operations.")
+    endif()
 endif()
+cmake_pop_check_state()
 
 INCLUDE(TestBigEndian)
 TEST_BIG_ENDIAN(BIGENDIAN)
```

**Comment by amak79** [comment]
https://github.com/badaix/snapcast/commit/64c78d1a1401478dcc49b3094dc779d4b08ec51e fails with the following error:
```
-- The CXX compiler identification is GNU 9.2.0
-- Check for working CXX compiler: /builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++
-- Check for working CXX compiler: /builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is GNU 9.2.0
-- Check for working C compiler: /builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-gcc
-- Check for working C compiler: /builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test HAVE_CXX_ATOMICS64_WITHOUT_LIB
-- Performing Test HAVE_CXX_ATOMICS64_WITHOUT_LIB - Failed
-- Looking for __atomic_load_8 in atomic
-- Looking for __atomic_load_8 in atomic - not found
-- Performing Test HAVE_CXX_LIBATOMICS64_2ARGS
-- Performing Test HAVE_CXX_LIBATOMICS64_2ARGS - Failed
CMake Error at cmake/CheckAtomic.cmake:150 (message):
  Host compiler appears to require libatomic, but cannot find it.
Call Stack (most recent call first):
  CMakeLists.txt:80 (include)


-- Configuring incomplete, errors occurred!
See also "/builddir/snapcast-0.16.0/build/CMakeFiles/CMakeOutput.log".
See also "/builddir/snapcast-0.16.0/build/CMakeFiles/CMakeError.log".
```
CMakeError.log:
```
Performing C++ SOURCE FILE Test HAVE_CXX_ATOMICS64_WITHOUT_LIB failed with the following output:
Change Dir: /builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_2a42e/fast && /usr/bin/make -f CMakeFiles/cmTC_2a42e.dir/build.make CMakeFiles/cmTC_2a42e.dir/build
make[1]: Entering directory '/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_2a42e.dir/src.cxx.o
/builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++    -fstack-clash-protection -D_FORTIFY_SOURCE=2 -O2 -pipe -march=armv6 -mfpu=vfp -mfloat-abi=hard   -I/usr/arm-linux-gnueabihf/usr/include -DHAVE_CXX_ATOMICS64_WITHOUT_LIB -std=c++11   -o CMakeFiles/cmTC_2a42e.dir/src.cxx.o -c /builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp/src.cxx
Linking CXX executable cmTC_2a42e
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_2a42e.dir/link.txt --verbose=1
/builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++  -fstack-clash-protection -D_FORTIFY_SOURCE=2 -O2 -pipe -march=armv6 -mfpu=vfp -mfloat-abi=hard   -I/usr/arm-linux-gnueabihf/usr/include -DHAVE_CXX_ATOMICS64_WITHOUT_LIB -std=c++11  -Wl,-z,relro -Wl,-z,now -Wl,--as-needed    -L/usr/arm-linux-gnueabihf/usr/lib  -rdynamic CMakeFiles/cmTC_2a42e.dir/src.cxx.o  -o cmTC_2a42e 
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: CMakeFiles/cmTC_2a42e.dir/src.cxx.o: in function `main':
src.cxx:(.text.startup+0x1c): undefined reference to `__atomic_load_8'
/usr/lib/gcc/arm-linux-gnueabihf/9.2.0/../../../../arm-linux-gnueabihf/bin/ld: src.cxx:(.text.startup+0x28): undefined reference to `__atomic_is_lock_free'
collect2: error: ld returned 1 exit status
make[1]: *** [CMakeFiles/cmTC_2a42e.dir/build.make:87: cmTC_2a42e] Error 1
make[1]: Leaving directory '/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_2a42e/fast] Error 2


Source file was:

#include <atomic>
#include <cstdint>
std::atomic<uint64_t> x (0);
int main() {
  uint64_t i = x.load(std::memory_order_relaxed);
  return std::atomic_is_lock_free(&x);
}

Determining if the function __atomic_load_8 exists in the atomic failed with the following output:
Change Dir: /builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_02579/fast && /usr/bin/make -f CMakeFiles/cmTC_02579.dir/build.make CMakeFiles/cmTC_02579.dir/build
make[1]: Entering directory '/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_02579.dir/CheckFunctionExists.cxx.o
/builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++    -fstack-clash-protection -D_FORTIFY_SOURCE=2 -O2 -pipe -march=armv6 -mfpu=vfp -mfloat-abi=hard   -I/usr/arm-linux-gnueabihf/usr/include -DCHECK_FUNCTION_EXISTS=__atomic_load_8   -o CMakeFiles/cmTC_02579.dir/CheckFunctionExists.cxx.o -c /builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx
<command-line>: error: new declaration 'char __atomic_load_8()' ambiguates built-in declaration 'long long unsigned int __atomic_load_8(const volatile void*, int)' [-fpermissive]
<command-line>: note: in definition of macro 'CHECK_FUNCTION_EXISTS'
/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx: In function 'int main(int, char**)':
/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CheckLibraryExists/CheckFunctionExists.cxx:17:25: error: too few arguments to function 'long long unsigned int __atomic_load_8(const volatile void*, int)'
   17 |   CHECK_FUNCTION_EXISTS();
      |                         ^
make[1]: *** [CMakeFiles/cmTC_02579.dir/build.make:66: CMakeFiles/cmTC_02579.dir/CheckFunctionExists.cxx.o] Error 1
make[1]: Leaving directory '/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_02579/fast] Error 2



Performing C++ SOURCE FILE Test HAVE_CXX_LIBATOMICS64_2ARGS failed with the following output:
Change Dir: /builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_7d881/fast && /usr/bin/make -f CMakeFiles/cmTC_7d881.dir/build.make CMakeFiles/cmTC_7d881.dir/build
make[1]: Entering directory '/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_7d881.dir/src.cxx.o
/builddir/.xbps-snapcast/wrappers/arm-linux-gnueabihf-c++    -fstack-clash-protection -D_FORTIFY_SOURCE=2 -O2 -pipe -march=armv6 -mfpu=vfp -mfloat-abi=hard   -I/usr/arm-linux-gnueabihf/usr/include -DHAVE_CXX_LIBATOMICS64_2ARGS  -latomic   -o CMakeFiles/cmTC_7d881.dir/src.cxx.o -c /builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp/src.cxx
/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp/src.cxx: In function 'int main()':
/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp/src.cxx:3:27: error: incorrect number of arguments to function '__atomic_load'
    3 |   __atomic_load(nullptr, 0);
      |                           ^
make[1]: *** [CMakeFiles/cmTC_7d881.dir/build.make:66: CMakeFiles/cmTC_7d881.dir/src.cxx.o] Error 1
make[1]: Leaving directory '/builddir/snapcast-64c78d1a1401478dcc49b3094dc779d4b08ec51e/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:121: cmTC_7d881/fast] Error 2


Source file was:

int main() {
  __atomic_load(nullptr, 0);
  return 0;
}
```
https://github.com/badaix/snapcast/commit/7d8e77cfc0eea067703d77cc13a709d6af20f2aa fails during linking with `undefined reference to __atomic_store_8` error. The atomic operations checks succeed but they don't look right for armv6l:
```
-- Performing Test HAVE_CXX_ATOMICS_WITHOUT_LIB
-- Performing Test HAVE_CXX_ATOMICS_WITHOUT_LIB - Success
```

**Comment by badaix** [comment]
sorry, I've allready noticed yesterday that 64c78d1a1401478dcc49b3094dc779d4b08ec51e fails in the CI builds, and made an update 7d8e77cfc0eea067703d77cc13a709d6af20f2aa, that at least fixed the CI issue.

**Comment by xabolcs** [comment]
I created a few commits to incorporate the hints from @amak79 

Compare view: [develop ... my branch](https://github.com/badaix/snapcast/compare/017988f99eac715615a46440578e44a751fd7602...xabolcs:issue-490-checkatomic-cmake-and-the-atomic-store-8?expand=1).

**Comment by amak79** [comment]
@xabolcs, with your changes the build completes successfully.
```
-- Performing Test HAVE_CXX_ATOMICS_WITHOUT_LIB
-- Performing Test HAVE_CXX_ATOMICS_WITHOUT_LIB - Failed
-- Looking for __atomic_fetch_add_4 in atomic
-- Looking for __atomic_fetch_add_4 in atomic - not found
-- Performing Test HAVE_LIBATOMIC_2ARGS
-- Performing Test HAVE_LIBATOMIC_2ARGS - Failed
-- Performing Test HAVE_CXX_ATOMICS64_WITHOUT_LIB
-- Performing Test HAVE_CXX_ATOMICS64_WITHOUT_LIB - Failed
-- Looking for __atomic_load_8 in atomic
-- Looking for __atomic_load_8 in atomic - not found
-- Performing Test HAVE_CXX_LIBATOMICS64_2ARGS
-- Performing Test HAVE_CXX_LIBATOMICS64_2ARGS - Success
-- Performing Test HAVE_CXX_ATOMICS64_WITH_LIB
-- Performing Test HAVE_CXX_ATOMICS64_WITH_LIB - Success
```

**Comment by xabolcs** [comment]
Compile-tested with toolchains:
* `toolchain-aarch64_cortex-a72_gcc-8.3.0_musl`
* `toolchain-arm_cortex-a7+neon-vfpv4_gcc-8.3.0_musl_eabi`
* `toolchain-i386_pentium_gcc-8.3.0_musl`
* `toolchain-mips_24kc_gcc-8.3.0_musl`
* `toolchain-mipsel_24kc_gcc-8.3.0_musl`

**Comment by xabolcs** [comment]
Opened pull request against *develop* as it has already a related commit.

**Comment by xabolcs** [comment]
@amak79 the fix landed on **develop** :+1:

**Comment by amak79** [comment]
@xabolcs Thank you.

I tested develop d198b4d0b15215cf3841fbd09fa2f8e07ed3db49. Atomic detection appears to be working correctly, and all builds complete successfully.

One minor issue is that cmake (I'm using 3.15.5) now displays the following warning:
```
-- Looking for C++ include sys/types.h
CMake Warning (dev) at /usr/share/cmake-3.15/Modules/CheckIncludeFileCXX.cmake:79 (message):
  Policy CMP0075 is not set: Include file check macros honor
  CMAKE_REQUIRED_LIBRARIES.  Run "cmake --help-policy CMP0075" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  CMAKE_REQUIRED_LIBRARIES is set to:

    atomic

  For compatibility with CMake 3.11 and below this check is ignoring it.
Call Stack (most recent call first):
  /usr/share/cmake-3.15/Modules/CheckTypeSize.cmake:234 (check_include_file_cxx)
  /usr/share/cmake-3.15/Modules/TestBigEndian.cmake:33 (CHECK_TYPE_SIZE)
  CMakeLists.txt:84 (TEST_BIG_ENDIAN)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Looking for C++ include sys/types.h - found
```
I had the same warning in https://github.com/badaix/snapcast/issues/490#issuecomment-546874663 but I suppressed it with `cmake_push_check_state()` / `cmake_pop_check_state()`. I don't know if that's the correct way to do it, and I was using `set(CMAKE_REQUIRED_LIBRARIES atomic)` instead of `list(APPEND CMAKE_REQUIRED_LIBRARIES "atomic")`.

**Comment by badaix** [comment]
should be fixed in 0.17
